feat: calendar core

bart-krakowski opened the pull request:
## Calendar Core
### Summary
The CalendarCore class provides a set of functionalities for managing calendar events, view modes, and period navigation. This class is designed to be used in various calendar applications where precise date management and event handling are required.

### Motivation
The motivation behind this class is to provide a reusable and efficient core for calendar functionalities that can be integrated into different applications, ensuring consistent and simplified management of calendar events, view modes, and period navigation.

### Parameters

- `weekStartsOn?: number`
  - An optional number that specifies the day of the week that the calendar should start on. It defaults to 1 (Monday).
- `events?: TEvent[]`
  - An optional array of events that the calendar should display.
- `viewMode: ViewMode`
  - An object that specifies the initial view mode of the calendar.
- `locale?: Parameters<Temporal.PlainDate['toLocaleString']>['0']`
  - An optional string that specifies the locale to use for formatting dates and times.


### Returns

- `getDaysWithEvents(): Array<Day<TEvent>>`
  - Returns an array of days in the current period with their associated events.
- `getDaysNames(): string[]`
  - Returns an array of strings representing the names of the days of the week based on the locale and week start day.
- `changeViewMode(newViewMode: ViewMode): void`
  - Changes the view mode of the calendar.
- `goToPreviousPeriod(): void`
  - Navigates to the previous period based on the current view mode.
- `goToNextPeriod(): void`
  - Navigates to the next period based on the current view mode.
- `goToCurrentPeriod(): void`
  - Navigates to the current period.
- `goToSpecificPeriod(date: Temporal.PlainDate): void`
  - Navigates to a specific period based on the provided date.
- `updateCurrentTime(): void`
  - Updates the current time.
- `getEventProps(id: Event['id']): { style: CSSProperties } | null`
  - Retrieves the style properties for a specific event based on its ID.
- `getCurrentTimeMarkerProps(): { style: CSSProperties; currentTime: string | undefined }`
  - Retrieves the style properties and current time for the current time marker.
- `groupDaysBy(props: Omit<GroupDaysByProps<TEvent>, 'weekStartsOn'>): (Day<TEvent> | null)[][]`
  - Groups the days in the current period by a specified unit. The fillMissingDays parameter can be used to fill in missing days with previous or next month's days.



### Example Usage
```ts
import { CalendarCore, Event } from '@tanstack/time';
import { Temporal } from '@js-temporal/polyfill';

interface MyEvent extends Event {
  location: string;
}

const events: MyEvent[] = [
  {
    id: '1',
    startDate: Temporal.PlainDateTime.from('2024-06-10T09:00'),
    endDate: Temporal.PlainDateTime.from('2024-06-10T10:00'),
    title: 'Event 1',
    location: 'Room 101',
  },
  {
    id: '2',
    startDate: Temporal.PlainDateTime.from('2024-06-12T11:00'),
    endDate: Temporal.PlainDateTime.from('2024-06-12T12:00'),
    title: 'Event 2',
    location: 'Room 202',
  },
];

const calendarCore = new CalendarCore<MyEvent>({
  weekStartsOn: 1,
  viewMode: { value: 1, unit: 'month' },
  events,
  locale: 'en-US',
});

// Get days with events
const daysWithEvents = calendarCore.getDaysWithEvents();
console.log(daysWithEvents);

// Change view mode to week
calendarCore.changeViewMode({ value: 1, unit: 'week' });

// Navigate to the next period
calendarCore.goToNextPeriod();

// Update current time
calendarCore.updateCurrentTime();

// Get event properties
const eventProps = calendarCore.getEventProps('1');
console.log(eventProps);

// Get current time marker properties
const currentTimeMarkerProps = calendarCore.getCurrentTimeMarkerProps();
console.log(currentTimeMarkerProps);

// Group days by week
const groupedDays = calendarCore.groupDaysBy({ days: daysWithEvents, unit: 'week' });
console.log(groupedDays);
```


## useCalendar hook
### Summary
The useCalendar hook provides a comprehensive set of functionalities for managing calendar events, view modes, and period navigation.

### Motivation
The motivation behind this hook is to provide a reusable and efficient way to integrate calendar functionalities into React components, ensuring consistent and simplified management of calendar events, view modes, and period navigation.

### Parameters

- `weekStartsOn?: number`
  - This parameter is an optional number that specifies the day of the week that the calendar should start on. It defaults to 0, which is Sunday.
- `events: Event[]`
  - This parameter is an array of events that the calendar should display.
- `viewMode: 'month' | 'week' | number`
  - This parameter is a string that specifies the initial view mode of the calendar. It can be either 'month', 'week', or a number representing the number of days in a custom view mode.
- `locale?: string`
  - This parameter is an optional string that specifies the locale to use for formatting dates and times. It defaults to the system locale.
- `onChangeViewMode?: ({ value: number; unit: "month" | "week" | "day"; }) => void`
  - This parameter is an optional callback function that is called when the view mode of the calendar changes. It receives the new view mode as an argument.
- `onChangeViewMode?: (viewMode: value: number; unit: "month" | "week" | "day";) => void`
  - This parameter is an optional callback function that is called when the view mode of the calendar changes. It receives the new view mode as an argument.
- `reducer?: (state: CalendarState, action: CalendarAction) => CalendarState`
  - This parameter is an optional custom reducer function that can be used to manage the state of the calendar.


### Returns

- `firstDayOfPeriod: Temporal.PlainDate`
  - This value represents the first day of the current period displayed by the calendar.
- `currentPeriod: string`
  - This value represents a string that describes the current period displayed by the calendar.
- `goToPreviousPeriod: MouseEventHandler<HTMLButtonElement>`
  - This function is a click event handler that navigates to the previous period.
- `goToNextPeriod: MouseEventHandler<HTMLButtonElement>`
  - This function is a click event handler that navigates to the next period.
- `goToCurrentPeriod: MouseEventHandler<HTMLButtonElement>`
  - This function is a click event handler that navigates to the current period.
- `goToSpecificPeriod: (date: Temporal.PlainDate) => void`
  - This function is a callback function that is called when a date is selected in the calendar. It receives the selected date as an argument.
- `days: Day[]`
  - This value represents an array of days in the current period displayed by the calendar.
- `daysNames: string[]`
  - This value represents an array of strings that contain the names of the days of the week.
- `viewMode: 'month' | 'week' | number`
  - This value represents the current view mode of the calendar.
- `changeViewMode: (newViewMode: 'month' | 'week' | number) => void`
  - This function is used to change the view mode of the calendar.
- `getEventProps: (id: string) => { style: CSSProperties } | null`
  - This function is used to retrieve the style properties for a specific event based on its ID.
- `getEventProps: (id: string) => { style: CSSProperties } | null`
  - This function is used to retrieve the style properties for a specific event based on its ID.
- `getEventProps: (id: string) => { style: CSSProperties } | null`
  - This function is used to retrieve the style properties for a specific event based on its ID.
- `getCurrentTimeMarkerProps: () => { style: CSSProperties, currentTime: Temporal.PlainTime }`
  - This function is used to retrieve the style properties and current time for the current time marker.
- `isPending: boolean`
  - This value represents whether the calendar is in a pending state.
- `groupDaysBy: ({ days: Day[], unit: 'week' | 'month', fillMissingDays?: boolean }) => Day[][]`
  - This function is used to group the days in the current period by a specified unit. The `fillMissingDays` parameter can be used to fill in missing days with previous or next month's days.

### Example Usage

```tsx
const CalendarComponent = ({ events }) => {
  const {
    firstDayOfPeriod,
    currentPeriod,
    goToPreviousPeriod,
    goToNextPeriod,
    goToCurrentPeriod,
    goToSpecificPeriod,
    changeViewMode,
    days,
    daysNames,
    viewMode,
    getEventProps,
    getCurrentTimeMarkerProps,
    groupDaysBy,
  } = useCalendar({
    weekStartsOn: 1,
    viewMode: { value: 1, unit: 'month' },
    locale: 'en-US',
    onChangeViewMode: (newViewMode) => console.log('View mode changed:', newViewMode),
  });

  return (
    <div className="calendar-container">
      <div className="calendar-header">
        <button onClick={goToPreviousPeriod}>Previous</button>
        <button onClick={goToCurrentPeriod}>Today</button>
        <button onClick={goToNextPeriod}>Next</button>
      </div>
      <div className="calendar-view-mode">
        <button onClick={() => changeViewMode({ value: 1, unit: 'month' })}>Month View</button>
        <button onClick={() => changeViewMode({ value: 1, unit: 'week' })}>Week View</button>
        <button onClick={() => changeViewMode({ value: 3, unit: 'day' })}>3-Day View</button>
        <button onClick={() => changeViewMode({ value: 1, unit: 'day' })}>1-Day View</button>
      </div>
      <table className="calendar-table">
        {viewMode.unit === 'month' && (
          groupDaysBy(days, 'months').map((month, monthIndex) => (
            <tbody key={monthIndex} className="calendar-month">
              <tr>
                <th colSpan={7} className="calendar-month-name">
                  {month[0]?.date.toLocaleString('default', { month: 'long' })}{' '}
                  {month[0]?.date.year}
                </th>
              </tr>
              <tr>
                {daysNames.map((dayName, index) => (
                  <th key={index} className="calendar-day-name">
                    {dayName}
                  </th>
                ))}
              </tr>
              {groupDaysBy(month, 'weeks').map((week, weekIndex) => (
                <tr key={weekIndex} className="calendar-week">
                  {week.map((day) => (
                    <td
                      key={day.date.toString()}
                      className={`calendar-day ${day.isToday ? 'today' : ''
                        } ${day.isInCurrentPeriod ? 'current' : 'not-current'}`}
                    >
                      <div className="calendar-date">{day.date.day}</div>
                      <div className="calendar-events">
                        {day.events.map((event) => (
                          <div
                            key={event.id}
                            className="calendar-event"
                            style={getEventProps(event.id)?.style}
                          >
                            {event.title}
                          </div>
                        ))}
                      </div>
                    </td>
                  ))}
                </tr>
              ))}
            </tbody>
          ))
        )}

        {viewMode.unit === 'week' && (
          <tbody className="calendar-week">
            <tr>
              {daysNames.map((dayName, index) => (
                <th key={index} className="calendar-day-name">
                  {dayName}
                </th>
              ))}
            </tr>
            {groupDaysBy(days, 'weeks').map((week, weekIndex) => (
              <tr key={weekIndex}>
                {week.map((day) => (
                  <td
                    key={day.date.toString()}
                    className={`calendar-day ${day.isToday ? 'today' : ''
                      } ${day.isInCurrentPeriod ? 'current' : 'not-current'}`}
                  >
                    <div className="calendar-date">{day.date.day}</div>
                    <div className="calendar-events">
                      {day.events.map((event) => (
                        <div
                          key={event.id}
                          className="calendar-event"
                          style={getEventProps(event.id)?.style}
                        >
                          {event.title}
                        </div>
                      ))}
                    </div>
                  </td>
                ))}
              </tr>
            ))}
          </tbody>
        )}

        {viewMode.unit === 'day' && (
          <tbody className="calendar-day">
            <tr>
              {daysNames.map((dayName, index) => (
                <th key={index} className="calendar-day-name">
                  {dayName}
                </th>
              ))}
            </tr>
            <tr>
              {days.map((day) => (
                <td
                  key={day.date.toString()}
                  className={`calendar-day ${day.isToday ? 'today' : ''
                    } ${day.isInCurrentPeriod ? 'current' : 'not-current'}`}
                >
                  <div className="calendar-date">{day.date.day}</div>
                  <div className="calendar-events">
                    {day.events.map((event) => (
                      <div
                        key={event.id}
                        className="calendar-event"
                        style={getEventProps(event.id)?.style}
                      >
                        {event.title}
                      </div>
                    ))}
                  </div>
                </td>
              ))}
            </tr>
          </tbody>
        )}
      </table>
    </div>
  );
};
```

bart-krakowski committed:
refactor: update isValidDate return type for improved type safety

bart-krakowski committed:
refactor: update isValidDate return type for improved type safety

bart-krakowski committed:
refactor: update isValidDate return type for improved type safety

bart-krakowski committed:
test: refactor tests for the parse util

bart-krakowski committed:
refactor: types for utils

bart-krakowski committed:
revert: isValidDate util

bart-krakowski committed:
Merge branch 'v0' into chore/vitest-config

bart-krakowski committed:
docs: architecture md file

bart-krakowski committed:
feat: useCalendar hook

bart-krakowski committed:
test: useCalendar tests

bart-krakowski committed:
docs: useCalendar hook

bart-krakowski committed:
docs: useCalendar hook

bart-krakowski committed:
test: useCalendar tests

bart-krakowski committed:
test: useCalendar tests

bart-krakowski committed:
test: overlaping events

bart-krakowski committed:
docs: add the choosing-a-date-time-handling-solution adr

bart-krakowski committed:
feat(react-time/useCalendar): getCurrentTimeMarkerProps

bart-krakowski committed:
feat(react-time/useCalendar): getCurrentTimeMarkerProps

bart-krakowski committed:
feat(react-time/useCalendar): getCurrentTimeMarkerProps

bart-krakowski committed:
refactor: reducer

bart-krakowski committed:
refactor: reducer

bart-krakowski committed:
refactor(useCalendar): types

bart-krakowski committed:
refactor(useCalendar): types

bart-krakowski committed:
refactor: useCalendar reducer

bart-krakowski committed:
refactor: useCalendar reducer

bart-krakowski committed:
refactor: useCalendar reducer

bart-krakowski committed:
feat: add the isToday field

bart-krakowski committed:
feat: add the isToday field

bart-krakowski committed:
feat: add the isToday field

bart-krakowski committed:
docs: useCalendar

bart-krakowski committed:
feat: useCalendar hook

bart-krakowski committed:
cd

bart-krakowski committed:
feat: useCalendar hook

bart-krakowski committed:
feat: useCalendar hook

bart-krakowski committed:
feat: useCalendar hook

bart-krakowski committed:
feat: useCalendar hook

bart-krakowski committed:
test: useCalendar

bart-krakowski committed:
test: useCalendar

bart-krakowski committed:
test: useCalendar

bart-krakowski committed:
feat: custom reducer

bart-krakowski committed:
docs: tsdocs

bart-krakowski committed:
docs: tsdocs

bart-krakowski committed:
docs: tsdocs

bart-krakowski committed:
refactor: useCalendar

bart-krakowski committed:
refactor: useCalendar

bart-krakowski committed:
refactor: useCalendar

bart-krakowski committed:
refactor: change viewMode type

bart-krakowski committed:
refactor: change viewMode type

bart-krakowski committed:
fix: missing days in view

bart-krakowski committed:
refactor: api

bart-krakowski committed:
refactor: api

bart-krakowski committed:
fix: types

bart-krakowski committed:
fix: types

bart-krakowski committed:
fix: types

bart-krakowski committed:
fix: types

bart-krakowski committed:
docs: useCalendar

bart-krakowski committed:
docs: useCalendar

bart-krakowski committed:
fix: types

bart-krakowski committed:
refactor: add the fillMissingDays argument

bart-krakowski committed:
refactor: add the fillMissingDays argument

bart-krakowski committed:
refactor: add the fillMissingDays argument

bart-krakowski committed:
refactor: move utilities to the core package

bart-krakowski committed:
fix: getCurrentTimeMarkerProps

bart-krakowski committed:
docs: update jsdocs

bart-krakowski committed:
test: useCalendar

bart-krakowski committed:
refactor: getCurrentTimeMarkerProps

bart-krakowski committed:
feat: calendar core

bart-krakowski committed:
feat: calendar core

bart-krakowski committed:
feat: calendar core

bart-krakowski committed:
docs: calendar core

bart-krakowski committed:
refactor: useCalendar

bart-krakowski committed:
test: calendar core

bart-krakowski committed:
refactor: useCalendar

bart-krakowski committed:
refactor: useCalendar

bart-krakowski committed:
refactor: useCalendar

bart-krakowski committed:
refactor(useCalendar): types

bart-krakowski committed:
refactor: move calendar-related files to the core directory

bart-krakowski committed:
refactor: Update CalendarCoreOptions interface in calendar.ts

bart-krakowski committed:
test: imports

bart-krakowski committed:
docs: typo

bart-krakowski committed:
refactor: DatePickerCore

bart-krakowski committed:
refactor: weekInfoPolyfill

bart-krakowski committed:
refactor: weekInfoPolyfill

bart-krakowski committed:
refactor: weekInfoPolyfill

bart-krakowski committed:
refactor: weekInfoPolyfill

bart-krakowski committed:
refactor: date defaults

bart-krakowski committed:
refactor: date defaults

bart-krakowski committed:
refactor: date defaults

bart-krakowski committed:
refactor: types

bart-krakowski committed:
docs: jsdocs

bart-krakowski committed:
refactor: types

bart-krakowski committed:
refactor: types

bart-krakowski committed:
test: useCalendar

bart-krakowski committed:
refactor: useIsomorphicLayoutEffect

bart-krakowski committed:
docs: calendar

bart-krakowski committed:
docs: calendar

cutterbl commented on packages/time/src/calendar/generateDateRange.ts:
```diff
@@ -0,0 +1,14 @@
+import { Temporal } from '@js-temporal/polyfill'
+
+export const generateDateRange = (
+  start: Temporal.PlainDate,
```
Method should take date time string and convert, or PlainDate. Remember [The Case for 'Simple' Dates](https://github.com/TanStack/time/wiki/TanStack-Time#the-case-for-simple-dates) from the wiki

bart-krakowski replied:
Sure! Do we want to require a specific format from the user? I'm wondering how to ensure type safety here

cutterbl replied:
I [wrote something up](https://github.com/TanStack/time/blob/v0/packages/time/src/utils/parsingDates.md) on this. This is the reasoning behind the `validateDate` in the utils, where a Date must be a `Date`, `string` (RFC 3339) or `number` (epoch). Simple types are easier to memoize against, whereas Date or Temporal objects are not.

cutterbl commented on packages/time/src/calendar/getEventProps.ts:
```diff
@@ -0,0 +1,78 @@
+  const event = [...eventMap.values()].flat().find((currEvent) => currEvent.id === id);
+  if (!event) return null;
+
+  const eventStartDate = Temporal.PlainDateTime.from(event.startDate);
```
Full date time should always be represented as ZoneDateTime

cutterbl commented on packages/time/src/calendar/getEventProps.ts:
```diff
@@ -0,0 +1,78 @@
+    if (isStartPart) {
+      const eventTimeInMinutes = eventStartDate.hour * 60 + eventStartDate.minute;
+      percentageOfDay = (eventTimeInMinutes / (24 * 60)) * 100;
+      eventHeightInMinutes = 24 * 60 - eventTimeInMinutes;
```
Not sure what this is. Can you explain it to me?

cutterbl commented on packages/time/src/calendar/getEventProps.ts:
```diff
@@ -0,0 +1,78 @@
+    eventHeightInMinutes = endTimeInMinutes - eventTimeInMinutes;
+  }
+
+  const eventHeight = Math.min((eventHeightInMinutes / (24 * 60)) * 100, 20);
```
Again, what is this?

cutterbl commented on packages/time/src/calendar/getEventProps.ts:
```diff
@@ -0,0 +1,78 @@
+  const eventLeft = sidePadding + eventIndex * (eventWidth + innerPadding);
+
+  if (state.viewMode.unit === 'week' || state.viewMode.unit === 'day') {
+    return {
```
Not sure of all of this. `@tanstack/time` won't really do layout, so much as deal with data and state. Layout is on the developer, who might use flexbox, CSS grid, or even tables. This might be worth some discussion.

bart-krakowski replied:
Totally! I'm gonna remove all the layout-specific code and replace it with more generic properties.

cutterbl commented on packages/time/src/calendar/groupDaysBy.ts:
```diff
@@ -0,0 +1,97 @@
+
+export type GroupDaysByProps<TEvent extends Event = Event> = GroupDaysByMonthProps<TEvent> | GroupDaysByWeekProps<TEvent>;
+
+export const groupDaysBy = <TEvent extends Event>({
```
This is an interesting method. An Event has both a 'start' and an 'end', meaning an Event can span units. For any given unit the 'start' may exist in one 'week' or 'month', while the 'end' may exist in another. 

The logic is a little more complex for both unit scenarios, and may be better served as two separate methods, even if those methods are called within this method. 

Also I would suggest sorting the array of 'days' before looping. This is relatively easy with Temporal's 'compare()' methods.

bart-krakowski replied:
Yeah, this method is intended only to group, for example, years into months and months into weeks. I think we can create a separate method for generating the date range.

cutterbl commented on packages/time/src/calendar/splitMultiDayEvents.ts:
```diff
@@ -0,0 +1,23 @@
+
+  let currentDay = startDate;
+  while (Temporal.ZonedDateTime.compare(currentDay, endDate) < 0) {
+    const startOfCurrentDay = currentDay.with({ hour: 0, minute: 0, second: 0, millisecond: 0 });
```
We need 'startOf'/'endOf' helper methods

cutterbl commented on packages/time/src/calendar/types.ts:
```diff
@@ -0,0 +1,24 @@
+
+export interface Event {
+  id: string;
+  startDate: Temporal.PlainDateTime | Temporal.ZonedDateTime;
```
Go with 'start' and 'end', it's more concise. A name of 'startDate' implies it is only a 'date' rather than a 'date/time'.

cutterbl commented on packages/time/src/calendar/types.ts:
```diff
@@ -0,0 +1,24 @@
+import type { Temporal } from "@js-temporal/polyfill"
+
+export interface Event {
```
Consider adding a 'resources' prop. An array of Resource id's. An Event is often attributed to some (or multiple) thing(s): People, Rooms, etc. Consider a calendar used for assigning health care staff to a procedure. Little Timmy is getting a tonsilectomy, where it will occur in OR 2 (location), and require an anesthesiologist, surgeon, and a nurse in attendance. One calendar might assign the Event to the available room Resource (OR2 of 1,2,3 and 4), while another calendar might be used to apply the personnel Resources. A 'Resource' type should be generic and extensible, with an 'id' and a 'title', while the 'resources' prop should be an array of the Resource id's.

bart-krakowski replied:
Nice one! I never thought about that. 🤯

cutterbl commented on packages/time/src/core/calendar.ts:
```diff
@@ -0,0 +1,397 @@
+  /** The number of units for the view mode. */
+  value: number
+  /** The unit of time that the calendar view should display (month, week, or day). */
+  unit: 'month' | 'week' | 'day'
```
Also consider 'workweek', defaulted as "the days of the week not in the weekInfo.weekend", but can somehow be overridden.

cutterbl commented on packages/time/src/core/calendar.ts:
```diff
@@ -0,0 +1,397 @@
+ * of events, locale, time zone, and the calendar system.
+ * @template TEvent - Specifies the event type, extending a base Event type.
+ */
+export interface CalendarCoreOptions<TEvent extends Event = Event> {
```
I would include `resources?: TResource[] | null` as a list of possible associated 'resources'

cutterbl commented on packages/time/src/core/calendar.ts:
```diff
@@ -0,0 +1,397 @@
+  /** The current view mode of the calendar. */
+  viewMode: CalendarStore['viewMode']
+  /** The current date and time according to the calendar's time zone. */
+  currentTime: CalendarStore['currentTime']
```
But, is it? We don't want to update this value every second, as a variable. Better to provide a 'now()' method in state.

cutterbl commented on packages/time/src/core/calendar.ts:
```diff
@@ -0,0 +1,397 @@
+  }
+
+  private getCalendarDays() {
+    const start =
```
Another spot where 'startOf' helper methods would come in real handy

cutterbl commented on packages/time/src/core/calendar.ts:
```diff
@@ -0,0 +1,397 @@
+    let end
+    switch (this.store.state.viewMode.unit) {
+      case 'month': {
+        const lastDayOfMonth = this.getFirstDayOfMonth()
```
and 'endOf' helper methods

cutterbl commented on packages/time/src/core/calendar.ts:
```diff
@@ -0,0 +1,397 @@
+      months: this.store.state.viewMode.value - 1,
+    }).month
+
+    return allDays.filter(
```
Are you filtering here? Or sorting? You can use the Temporal 'compare()' methods as the sort method of an array

cutterbl commented on packages/time/src/core/calendar.ts:
```diff
@@ -0,0 +1,397 @@
+          this.options.timeZone,
+        )
+        splitEvents.forEach((splitEvent) => {
+          const splitKey = splitEvent.startDate.toString().split('T')[0]
```
This `const [ splitKey ] = splitEvent.startDate.toString().split('T')` will do the same thing, though I would call the variable `datePart` rather than `splitKey`. I know you're using it as the 'key' in the map, but I'd still call it what it is for clarity sake.

cutterbl commented on packages/time/src/core/calendar.ts:
```diff
@@ -0,0 +1,397 @@
+          }
+        })
+      } else {
+        const eventKey = event.startDate.toString().split('T')[0]
```
Same thoughts as last comment, though maybe `eventStartDate`.

cutterbl commented on packages/time/src/core/calendar.ts:
```diff
@@ -0,0 +1,397 @@
+ * @template TEvent - Specifies the event type, extending a base Event type.
+ */
+export interface CalendarCoreOptions<TEvent extends Event = Event> {
+  /** An optional array of events to be handled by the calendar. */
```
Calendars have an 'activeDate': the date which is the core reference date/time of a calendar display, and is generally highlighted during navigation. The 'activeDate' defaults to 'now', but it's possible this could be overridden. When moving back and forth within a 'unit' the `activeDate` (not to be confused with 'Today') would change by 'unit'. This is where Temporal.Duration definitions for 'month', 'week', and 'day' come in very handy, as they'll properly transition across ZonedDateTime instances, using 'add()' and 'subtract()' methods. So moving week to week, if the 'activeDate' is a Wednesday it stays a Wednesday, but moving to the previous month from July 31st should go to June 30th, but July 4th should go to June 4th.

cutterbl commented on packages/time/src/core/calendar.ts:
```diff
@@ -0,0 +1,397 @@
+  }
+
+  goToNextPeriod() {
+    const firstDayOfMonth = this.getFirstDayOfMonth()
```
This should change the 'activeDate' and all accompanying 'unit' specific data

cutterbl commented on packages/time/src/core/calendar.ts:
```diff
@@ -0,0 +1,397 @@
+  }
+
+  goToCurrentPeriod() {
+    this.store.setState((prev) => ({
```
This should change the 'activeDate' and all accompanying 'unit' specific data

cutterbl commented on packages/time/src/core/calendar.ts:
```diff
@@ -0,0 +1,397 @@
+  }
+
+  goToSpecificPeriod(date: Temporal.PlainDate) {
+    this.store.setState((prev) => ({
```
This should change the 'activeDate' and all accompanying 'unit' specific data

cutterbl commented on packages/time/src/core/calendar.ts:
```diff
@@ -0,0 +1,397 @@
+    const currentTimeInMinutes = hour * 60 + minute
+    const percentageOfDay = (currentTimeInMinutes / (24 * 60)) * 100
+
+    return {
```
This is also layout specific, so sort of depends on how the developer renders their calendar

cutterbl commented on packages/time/src/core/weekInfoPolyfill.ts:
```diff
@@ -0,0 +1,1864 @@
+}
+
+;(function () {
+  if (typeof (Intl as any).Locale.prototype.getWeekInfo !== 'function') {
```
Safari is the only browser (currently) with `Locale.prototype.getWeekInfo`. Chrome and Edge have a `weekInfo` accessor of `Locale`. Those browsers with the accessor should return it's value when calling `Local.prototype.getWeekInfo`, whereas those that do not (Firefox) would need the true polyfil method.

Since the polyfil is only applied to browsers that don't have `getWeekInfo`, and a true override method is only applied to one's without `Locale.weekInfo`, it would be good to dynamically load the additional data only to those browsers that require it.

cutterbl commented on packages/time/src/utils/getFirstDayOfWeek.ts:
```diff
@@ -0,0 +1,8 @@
+export const getFirstDayOfWeek = (currWeek: string, locale: Intl.UnicodeBCP47LocaleIdentifier | Intl.Locale = 'en-US') => {
+  const date = Temporal.PlainDate.from(currWeek);
+  const loc = new Intl.Locale(locale);
+  const { firstDay } = loc.weekInfo || loc.getWeekInfo();
```
With redesign of the polyfil you would only need to call `loc.getWeekInfo()` and it would work the same across all browsers.

cutterbl reviewed (changes requested):
Still figuring out nx and TS, so having trouble visualizing some of this. I've put down some notes on things, and hopefully they make sense. There is a lot here, so I'm sure I missed some stuff. 

A big thing is to consider the core idea laid out in the intro [wiki entry](https://github.com/TanStack/time/wiki/TanStack-Time) as well as the core idea around TanStack projects where we deal in data and state, and let the developer decide how they render the UI. Not to say that we won't provide any UI type props at all, but that we have to carefully think through what those things might be (likely ARIA type props and event handlers).

Please take all of this feedback only as constructive criticism. I really appreciate any assistance with this project. While I'm new to nx and TS, I've been a JS developer for 25+ years now, and have been working on scheduling applications for the better part of 20 years, so I'm speaking to you from years of experience in this area.

If you have questions on anything, please reach out.

bart-krakowski committed:
refactor: zonedDateTime

bart-krakowski committed:
refactor: rename startDate to start and endDate to end

bart-krakowski committed:
refactor: remove the getCurrentTimeMarkerProps method

bart-krakowski committed:
refactor: weekInfoPolyfill

bart-krakowski committed:
refactor: add a resources field to event

bart-krakowski committed:
refactor: add a resources field to event

bart-krakowski committed:
refactor: add a resources field to event

bart-krakowski committed:
refactor: add startOf and endOf helpers

bart-krakowski committed:
refactor: remove all the layout specific code

bart-krakowski committed:
refactor: update getDaysNames method to accept weekday parameter

bart-krakowski committed:
refactor: update useCalendar test to mock current date and time

bart-krakowski committed:
refactor: update useCalendar test to mock current date and time

bart-krakowski committed:
test: update useCalendar tests

bart-krakowski committed:
test: update useCalendar tests

bart-krakowski committed:
test: useCalendar

bart-krakowski committed:
refactor: update CalendarCore tests to mock current date and time

cutterbl commented on packages/time/src/core/weekInfoPolyfill/weekInfoPolyfill.ts:
```diff
@@ -0,0 +1,41 @@
+}
+
+;(function () {
+  // Chrome & Safari
```
Minor bit: This is for Chrome & Edge. (surprisingly) Safari provides `getWeekInfo()` natively. Really like this though. Looks good.

cutterbl commented on packages/time/src/core/weekInfoPolyfill/weekInfoPolyfill.ts:
```diff
@@ -0,0 +1,41 @@
+      Intl.Locale.prototype.getWeekInfo = function () {
+        const locale = this.toString().toLowerCase()
+
+        const match =
```
Two minor bits here: With your logic it would seem that you always have a `match`, so you shouldn't need the optional chaining. Also I would just use destructuring here `const { firstDay, weekend, minimalDays } = ...`, or return the value right here.

cutterbl commented on packages/time/src/utils/endOf.ts:
```diff
@@ -0,0 +1,15 @@
+/**
+ * Helper function to get the end of a given day.
+ * @param date - The date for which the end of the day is needed.
+ * @returns The end of the given day.
```
But is it the 'end of the given day'? Couldn't/shouldn't `endOf` be the end of any given unit? To get the 'end of the week' or 'end of the month', etc?

cutterbl commented on packages/time/src/utils/startOf.ts:
```diff
@@ -0,0 +1,17 @@
+/**
+ * Helper function to get the start of a given day.
+ * @param date - The date for which the start of the day is needed.
+ * @returns The start of the given day.
```
Same notes on `startOf` as I put above on `endOf`. Should really work for any given unit. Even 'work week', or even 'decade', as extension.

bart-krakowski committed:
refactor: Update CalendarCore to use destructuring assignment for datePart

bart-krakowski committed:
feat: add the activeDate property

bart-krakowski committed:
refactor: use compare method

bart-krakowski committed:
feat: add workWeek unit

bart-krakowski committed:
feat: add workWeek unit

bart-krakowski committed:
test: tests reorganisation

bart-krakowski committed:
chore: add the get-week-info-polyfill

bart-krakowski committed:
feat: getWeekInfo polyfill

bart-krakowski committed:
refactor: getEventProps

bart-krakowski committed:
refactor: getEventProps

bart-krakowski committed:
refactor: Update generateDateRange to accept string inputs

bart-krakowski committed:
chore: upgrade polyfill

bart-krakowski committed:
fix: getEventProps

bart-krakowski committed:
fix: getEventProps

bart-krakowski committed:
refactor: update date inputs to accept string format

bart-krakowski committed:
refactor: convert activeDate, currentPeriod to string in useCalendar

bart-krakowski committed:
refactor: startOf and endOf helpers

bart-krakowski committed:
test: calendar core

bart-krakowski committed:
refactor: update useCalendar to accept string inputs for activeDate and currentPeriod

bart-krakowski committed:
refactor: Update get-week-info-polyfill dependency to version 1.0.6

bart-krakowski committed:
refactor: Update startOf and endOf helpers to accept string inputs

cutterbl commented on packages/time/src/utils/endOf.ts:
```diff
@@ -1,67 +1,68 @@
  */
-export const endOf = (date: Temporal.ZonedDateTime, unit: 'day' | 'week' | 'month' | 'year' | 'workWeek' | 'decade'): Temporal.ZonedDateTime => {
+export function endOf({
+  date,
```
Like the direction here. I started a discussion in the time-maintainers discord around these types of methods...

cutterbl reviewed (commented):
See the discussion in the time-maintainers Discord channel

bart-krakowski committed:
docs: update architecture

bart-krakowski committed:
# This is a combination of 6 commits.
# This is the 1st commit message:

docs: readme

# This is the commit message #2:

docs: readme

# This is the commit message #3:

docs: update the repo header

# This is the commit message #4:

docs: update the repo header

# This is the commit message #5:

docs: update the repo header

# This is the commit message #6:

docs: update the repo header

bart-krakowski committed:
docs: readme

bart-krakowski committed:
chore: dependencies update

bart-krakowski committed:
chore: dependencies update

bart-krakowski committed:
Merge branch 'TanStack:v0' into v0

bart-krakowski committed:
chore: resolve conflicts

bart-krakowski committed:
Merge pull request #1 from bart-krakowski/chore/vitest-config

Chore/vitest config

bart-krakowski committed:
docs: architecture

docs: architecture

bart-krakowski committed:
Merge branch 'v0-1-0' into docs/readme

bart-krakowski committed:
docs: readme

docs: readme

bart-krakowski committed:
docs: adr choosing a date and time handling solution

docs: add the choosing-a-date-time-handling-solution adr

bart-krakowski committed:
fix: parse util

bart-krakowski committed:
fix: parse util

bart-krakowski committed:
test: parse

test: refactor tests for the parse util

bart-krakowski committed:
fix: parse util

bart-krakowski committed:
Merge branch 'v0-1-0' of github.com:bart-krakowski/time into v0-1-0

bart-krakowski committed:
chore: resolve conflicts

bart-krakowski committed:
chore: resolve conflicts

bart-krakowski committed:
chore: resolve conflicts

bart-krakowski committed:
test: update temporal mocks

bart-krakowski committed:
chore: update the lock file

bart-krakowski committed:
fix: lint

bart-krakowski committed:
fix: lint

bart-krakowski committed:
fix: lint

bart-krakowski committed:
feat: basic example

bart-krakowski committed:
feat: basic example

bart-krakowski committed:
feat: basic example

bart-krakowski committed:
feat: calendar

bart-krakowski committed:
feat: calendar

bart-krakowski committed:
feat: calendar

bart-krakowski committed:
feat: calendar

bart-krakowski committed:
feat: calendar

bart-krakowski committed:
feat: calendar
